src/index.rs

```diff
@@ -179,56 +181,59 @@ impl Index {
             .filter_map(move |height| self.chain.get_block_hash(height))
     }
 
-    pub(crate) fn sync(&mut self, daemon: &Daemon, exit_flag: &ExitFlag) -> Result<()> {
-        self.stats.observe_db(&self.store);
-        loop {
-            let new_headers = self
-                .stats
-                .observe_duration("headers", || daemon.get_new_headers(&self.chain))?;
-            if new_headers.is_empty() {
-                break;
-            }
-            info!(
-                "indexing {} blocks: [{}..{}]",
-                new_headers.len(),
-                new_headers.first().unwrap().height(),
-                new_headers.last().unwrap().height()
-            );
-            for chunk in new_headers.chunks(self.batch_size) {
-                exit_flag.poll().with_context(|| {
-                    format!(
-                        "indexing interrupted at height: {}",
-                        chunk.first().unwrap().height()
-                    )
-                })?;
-                let blockhashes: Vec<BlockHash> = chunk.iter().map(|h| h.hash()).collect();
-                let mut heights = chunk.iter().map(|h| h.height());
+    // Return `Ok(true)` when the chain is fully synced and the index is compacted.
+    pub(crate) fn sync(&mut self, daemon: &Daemon, exit_flag: &ExitFlag) -> Result<bool> {
+        let new_headers = self
+            .stats
+            .observe_duration("headers", || daemon.get_new_headers(&self.chain))?;
+        if new_headers.is_empty() {
+            self.store.flush(); // full compaction is performed on the first flush call
+            self.is_ready = true;
+            return Ok(true); // no more blocks to index (done for now)
+        }
+        info!(
+            "indexing {} blocks: [{}..{}]",
+            new_headers.len(),
+            new_headers.first().unwrap().height(),
+            new_headers.last().unwrap().height()
+        );
```

Comment:
If you'd like to avoid unwrap:

```rust
match (new_headers.first(), new_headers.last()) {
    (Some(first), Some(last)) => {
        info!(
            "indexing {} blocks: [{}..{}]",
            new_headers.len(),
            new_headers.first().unwrap().height(),
            new_headers.last().unwrap().height()
        );
    },
    _ => {
        self.store.flush(); // full compaction is performed on the first flush call
        self.is_ready = true;
        return Ok(true); // no more blocks to index (done for now)
    },
}
```

Reply:
-> https://github.com/romanz/electrs/pull/543/commits/2bfd974c72ea5c82a08e7cd709595bfe141943a0